docs/joss/paper.md

```diff
@@ -0,0 +1,193 @@
+and any non-default model components with its respective parameters.
+All components are then collected into a single model object, which, once
+initialized, returns a simulation object that contains the entire model state,
+work arrays and parameters, that can be run, analysed or changed.
+While these steps can be written into a script for reproducibility,
+the same steps can be executed and interacted with one-by-one in
+Julia's read-evaluate-print loop (REPL). We thereby reach an interactivity
+far beyond a monolithic interface that is limited to the options provided.
+
+To be extendible and composable with new
+model components, SpeedyWeather.jl relies on Julia's multiple dispatch
+programming paradigm [@Bezanson2017]. Every model component
+is defined as a new type. For example, to define a new way how to calculate
+the precipitation due to the physical process of large-scale condensation,
+one would define `MyCondensation` as a new subtype of `AbstractCondensation`.
+One then only needs to extend the `initialize!` and `condensation!`
+functions for this new type. Passing on `condensation = MyCondensation()`
+to the model constructor then implements this new model component without
+the need to branch off or overwrite existing model components.
+Conceptually similar scientific modelling paradigms have been very successful
+in the Python-based generic partial differential equation solver Dedalus [@Burns2020]
+and the Julia ocean model Oceananigans.jl [@Ramadhan2020].
+
+![Surface temperature simulated with the primitive equation model in SpeedyWeather.jl.
+(Figure will be updated) \label{fig:primitive}](primitive.jpg)
+
+The dynamical core of SpeedyWeather.jl uses established numerics
+[@Bourke1972; @Hoskins1975; @Simmons1978; @Simmons1981],
+widely adapted in numerical weather prediction. It is based on the spherical
+harmonic transform with a leapfrog-based semi-implicit time integration [@Hoskins1975]
+and a Robert-Asselin-Williams filter [@Williams2011; @Amezcua2011].
+The spherical harmonic transform is grid-flexible. Any iso-latitude ring-based
+grid can be used and new grids can be externally defined and passed on
+as argument. Many grids are already implemented: The conventional
+Gaussian grid, a regular longitude-latitude grid, 
+the octahedral Gaussian grid [@Malardel2016], the octahedral
+Clenshaw-Curtis grid [@Hotta2018], and the HEALPix grid [@Gorski2005].
+Both SpeedyWeather.jl and its spherical harmonic transform are also
+number format-flexible. 32-bit single-precision floating-point numbers
+(Float32) are the default as adapted by other modelling efforts [@Vana2017, @Nakano2018],
+but Float64 and other custom number formats can be used with a single
+code basis. Julia will compile to the choice of the number format, the grid,
+and and other model components just-in-time. A simple parallelisation
+across vertical layers is supported with Julia's multi threading.
```

Comment:
Hm... OK, so I read here that it's not parallelisable (eg MPI-type). Perhaps we can have a comment about "low memory allocation" (if that's true) "which allows for this and that..."

How about GPUs -- no mention about that? What's the state on running on GPUs (I am also asking out of curiosity, don't imply that we necessarily need to comment on the paper).

Reply:
Low memory allocation like Oceananigans can do it is really tricky with spectral models. At least I am not aware of a way to achieve that. As the transforms are costly and contain global communication you don't really have a bunch of stencils that can be fused to one...

Reply:
Some initial structure for GPUs is there, some Adapt calls, but no single kernel has actually been written for GPU. Maximilian experimented with a few in a branch but it's not in main. In general, I'm still not sure where SpeedyWeather should try to go. Multi-GPU or even Multi-CPU seems like a lot of work (domain decomposition in grid-point space and spectral space plus global communication), so I'm wondering whether a good goal for this model would be to aim for single GPU.

Reply:
I see. Yeah perhaps single GPU is a good first goal. But also no need to mention anything  in the paper then given the current status of affairs.